### PR TITLE
[CodeCompletion] Fix an issue that causes an iterator to be invalidated while iterating

### PR DESCRIPTION
**Pull request by ahoppen**
In `SourceLookupCache::lookupVisibleDecls`, copy the top level values before iterating them.

If we have 'addinitstotoplevel' enabled, calling 'addConstructorCallsForType' can cause macros to get expanded, which can then cause new members to get added to 'TopLevelValues', invalidating the current iterator.

I have not been able to reduce this to a test case that doesn’t rely on the `Observation` module in the SDK but here is the test case with which I was able to reproduce the issue very reliably.

```swift
import Foundation
import Observation

@Observable class MyObject {}

extension MyObject {}

// RUN: ~/sbin/sourcekitd-test \
// RUN:   -req=complete.open \
// RUN:   -req-opts=addinitstotoplevel=1 \
// RUN:   -pos=8:1 \
// RUN:   %s \
// RUN:   -- \
// RUN:   %s \
// RUN:   -Xfrontend \
// RUN:   -load-plugin-library \
// RUN:   -Xfrontend \
// RUN:   /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/host/plugins/libObservationMacros.dylib \
// RUN:   -sdk \
// RUN:   /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk
```

rdar://109202157

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Review by bnbarham — APPROVED** [review]

File: lib/IDE/CompletionLookup.cpp [hunk 1]
```diff
@@ -3214,16 +3214,61 @@ void CompletionLookup::getTypeCompletionsInDeclContext(SourceLoc Loc,
...
+    ValueDecl *VD;
+    DeclVisibilityKind Reason;
+    DynamicLookupInfo DynamicLookupInfo;
+  };
+
+  std::vector<FoundDecl> FoundDecls;
+
+  void foundDecl(ValueDecl *VD, DeclVisibilityKind Reason,
+                 DynamicLookupInfo DynamicLookupInfo = {}) override {
+    FoundDecls.push_back({VD, Reason, DynamicLookupInfo});
```
Comment (bnbarham) [inline lib/IDE/CompletionLookup.cpp:3232]:
```suggestion
    FoundDecls.emplace_back(VD, Reason, DynamicLookupInfo);
```

File: lib/IDE/CompletionLookup.cpp [hunk 2]
```diff
@@ -3214,16 +3214,61 @@ void CompletionLookup::getTypeCompletionsInDeclContext(SourceLoc Loc,
...
   NeedLeadingDot = false;
 
+  // If we have 'addinitstotoplevel' enabled, calling `foundDecl` on `this`
+  // can cause macros to get expanded, which can then cause new members ot get
+  // added to 'TopLevelValues', invalidating iterator over `TopLevelDecls` in
+  // `SourceLookupCache::lookupVisibleDecls`.
+  //
+  // Technically `foundDecl` should not expand macros or discover new top level
+  // members in any way because those newly discovered decls will not be added
+  // to the code completion results. However, it's preferrable to miss results
```
Comment (bnbarham) [inline lib/IDE/CompletionLookup.cpp:3261]:
```suggestion
  // to the code completion results. However, it's preferable to miss results
```

**Review by rintaro — APPROVED** [review]

**Pull request merged**